db/compaction/compaction_picker_level.cc

```diff
@@ -448,6 +448,10 @@ bool LevelCompactionBuilder::PickFileToCompact() {
     // do not pick a file to compact if it is being compacted
     // from n-1 level.
     if (f->being_compacted) {
+      if (ioptions_.compaction_pri == kRoundRobin) {
+        // Advance the cursor if a file is being compacted
+        vstorage_->UpdateCompactCursor(start_level_, cmp_idx + 1);
```

Comment:
A few design questions:

- [ ] I think we should only update cursor when the compaction is done;

- [ ] for one level, only one compaction (the one starts from the cursor) is able to update the cursor when it's done, which can be blocked by:
1.  there's already another compaction starts from that level, or
2. a high level compaction, for example, in your "compacting" test case, L1 file `[100, 240]` is compacting L2 which overlaps with current cursor;
In both cases, the "next" file may likely also be overlapping with the ongoing compaction, in the following example, if `T1` is compacting to the next level, `T2` is the next file, which is unable to compact, because the next level file is compacting:
![image](https://user-images.githubusercontent.com/33968/173123668-985e6141-8adf-46c0-b03b-19c2c0ae8051.png)
So we could continue select the next next file T3, until we find one. Or another strategy is just fall back to `kMinOverlappingRatio`, because anyway these compaction won't update cursor (hopefully these compactions are only happening because of high compaction pressure)

- [ ] ideally, we should use subcompaction to increase the compaction parallelism instead of having multiple compactions, but deciding how many input files (subcompaction number) is challenge, as if we choose too many it may delay other levels compaction. We don't need to handle it for now, it will be the next step for the project.

Reply:
>  I think we should only update cursor when the compaction is done;

 I may have suggested the opposite to Zichen. Do we know the compaction reason at compaction done time? We only want to advance for `CompactionReason::kLevelMaxLevelSize` compactions not all compactions. This is known when we pick the compaction and it sounds like we don't plan to use the cursor again until after that compaction finishes regardless of where the update happens. The only difference then is whether the advance happens in case compaction fails. I am neutral on this matter so prefer whatever's simpler to implement. Note that LevelDB actually considered it a benefit to advance the cursor despite failure -- https://github.com/google/leveldb/blob/d019e3605f222ebc5a3a2484a2cb29db537551dd/db/version_set.cc#L1440-L1445.

Reply:
> This is known when we pick the compaction and it sounds like we don't plan to use the cursor again until after that compaction finishes regardless of where the update happens

Actually I'd take that back and guess it's OK to let picking try the next file. It'll almost certainly conflict due to output-level files in common with the concurrent compaction. But it doesn't seem bad to let it be picked if it somehow works out.